Ensure actioncable behaves as expected with non-string queues

iamvery opened the pull request:
# The problem

Using a non-string value for the queue name in ActionCable leads to baffling behavior. In my case, everything worked as expected locally (using the async adapter). When I delivered the app to prod (using the redis adapter), I was able to see heartbeat messages, but none of the application messages arrived. There were no errors in the logs.

After some debugging, I realized that the problem was the datatype of the queue used in my messages.

```
ActionCable.server.broadcast(:foos, foo: :bar)
```

Perhaps using a symbol as a queue name doesn't make sense, but for whatever reason it was my default behavior.

The error was more apparent with the PostgreSQL adapter as it's call to [`escape_identifier`](https://github.com/rails/rails/blob/c57e7239a8b82957bcb07534cb7c1a3dcef71864/actioncable/lib/action_cable/subscription_adapter/postgresql.rb#L15) fails for non-string types.
# The fix

I saw two options.
1. Reject non-string values
2. Convert values to string

I tend to think the former might be more desirable, namely "tell the user that there is a problem" vs. "it's wrong, but try to make it right". In this case, the implementation was the former was a bit simpler, so I opted to propose that change for feedback. As always, I'm happen to restructure in any way needed to help out.

Thanks :heart:

stevenharman commented on actioncable/lib/action_cable/subscription_adapter/postgresql.rb:
```diff
@@ -12,16 +12,16 @@ def initialize(*)
       def broadcast(channel, payload)
         with_connection do |pg_conn|
-          pg_conn.exec("NOTIFY #{pg_conn.escape_identifier(channel)}, '#{pg_conn.escape_string(payload)}'")
+          pg_conn.exec("NOTIFY #{pg_conn.escape_identifier(channel.to_s)}, '#{pg_conn.escape_string(payload)}'")
```
What about using the implicit conversion, which will allow for string-like objects, while failing (loudly) for non-string-like things?

```ruby
String(channel)
```

iamvery replied:
Oh that's solid!

iamvery replied:
Hmm... looks like Ruby's `String()` method isn't using the implicit protocol...

```
pry> String(Object.new)
=> "#<Object:0x007fd9f05553f8>"
pry> Object.new.to_str
NoMethodError: undefined method `to_str' for #<Object:0x007fd9ef19f070>
from (pry):3:in `__pry__'
pry> Object.new.to_s
=> "#<Object:0x007fd9f0579fa0>"
```

iamvery replied:
http://ruby-doc.org/core-2.3.0/Kernel.html#method-i-String
> First tries to call its `to_str` method, then its `to_s` method.
But apparently it doesn't fail if implicit conversion fails.

iamvery replied:
I also looked into using the implicit conversion directly `to_str`. Symbol doesn't implement that, so if we want it to fail for very nearly non-string types, this will work, but I'll need to update the PR and test.

stevenharman replied:
Ugh, I forget that `String()` falls back to `#to_s`.

In this case, it seems desirable to at least give the option of being string like (via `#to_str`) here and then fall back to `#to_s` in the worst case.

iamvery replied:
I went ahead and push the `String()` change. :heart: @stevenharman

maclover7 commented:
Thanks for this PR! I think it might be easier to only support string names instead of symbols (symbols are not supported currently) - was there any reason you went with symbols instead of strings?

iamvery commented:
I suppose it was just out of habit. There wasn't any reason in particular _not_ to use a symbol. Maybe I was thinking about the channel as something of a "key" in the cable.

After discussion with @stevenharman, I do like the of supporting conversion rather than doing a type check. One use case (shamelessly stollen from steven) would be if your app formalized the concept of a `Channel` as an object, which supported implicit conversion to string in order to support using it with AC

```
class Channel
  attr_reader :name

  def initialize(name)
    @name = name
  end

  def to_str
    name
  end
end

#...
channel = Channel.new("things")
ActionCable.server.broadcast(channel, ...)
```

Of course you could explicitly convert the channel object to a string before passing it along, but this is exactly why the conversion protocols exist.

Either way, let me know how to fix this up for you :blush:

rafaelfranca commented:
r? @dhh what do you think about this?

iamvery commented:
Just pushed a fix for the test failures. I didn't catch evented redis.

matthewd commented:
If we're going to do this, we should probably change the callers, rather than the individual adapters: it's safe, because the adapter is not exposed user code.

iamvery commented:
That makes sense, @matthewd. I tried to figure out where this change would go, and I'm having some difficulty. I believe it would be in the [server broadcasting](https://github.com/rails/rails/blob/353782652b7dae7573f6480439519f227c41185a/actioncable/lib/action_cable/server/broadcasting.rb#L28) and the [channel streams](https://github.com/rails/rails/blob/353782652b7dae7573f6480439519f227c41185a/actioncable/lib/action_cable/channel/streams.rb#L72). If that's right, I'm not sure what would be the best way to test those situations... Am I on the right track?

maclover7 commented:
Personally, I'm :-1: on this, but I'll leave it up to you folks :)

iamvery commented:
Even if we're not on board with converting the input to string, I definitely feel like we should at the very least raise an error to limit the confusion. I spent a couple hours over a couple days of mine and some other's time trying to track the problem down. It's completely invisible to developer.

The experience from that side is
1. everything works as expected in development mode
2. the connect, heartbeat, etc. work in production, but messages sent from app appear to "disappear" without anything log output or anything

matthewd commented:
@iamvery https://github.com/rails/rails/blob/353782652b7dae7573f6480439519f227c41185a/actioncable/lib/action_cable/server/broadcasting.rb#L42, and two calls each of `pubsub.subscribe` / `pubsub.unsubscribe`.

Absent any simpler ideas, you should be able to test it by running up a full stack (a la `client_test.rb`), passing non-strings, and confirming we get the desired results.

iamvery commented:
Thanks for the pointer, @matthewd. I'm afraid we'll have to make more changes to the client test than is desirable. Currently it uses the async adapter which does not exhibit the invalid behavior. So as far as I can tell, the other adapters aren't covered by a full stack test. Can you confirm?

matthewd commented:
Correct. But if we're testing code that exists outside the adapters, do we care?

iamvery commented:
Yes, I believe we should care that there is no failing test for the case. That is, if I add channel that subscribes with a symbol to the client test, there would be no failure currently. Check out this patch: https://gist.github.com/iamvery/ef0aa4cc4e10fecfc598. You should see all the actioncable tests passing. If you swap out the config line for the commented on in the client test, you will see a failing test. :confused:

matthewd commented:
Right.. but that test isn't *using* the `stream_from`. If you then broadcast to `"symbol"` and assert it arrives at the client, we prove that stringification is occurring (because otherwise the symbol wouldn't match the string).

However, I think I'm leaning towards the stricter option: call `to_str`, and barf on anything else... optionally with a special case for Symbol. I think if we allow arbitrary objects, it'll be far more common for that to be an accidental error causing "why isn't my thing arriving?" confusion, than for it to be cleverly employed with a rich channel object.

dhh commented:
I think wanting to use a symbol for the channel name is a very natural inclination. I think that should Just Work. No need for us to be pedantic about that. So I say we convert to string. As long as it's clear in the log what's actually going on (i.e. which channel name is being used).

iamvery committed:
whitespace

iamvery committed:
Convert stream broadcasting to a string

ActionCable does some things behind the scenes that expects these
"broadcasting"s or "channel"s to be strings. However it's not
immediately obvious that the value must be a string. So adding this
conversion ensures things work as expected.

iamvery committed:
Ensure server broadcasts are to string queue names

Similar to the channel streaming side, these values must be strings for
ActionCable to behave as expected. The conversion will allow users to
send string-convertible values and get the expected behavior.

iamvery committed:
Update changelog

iamvery commented:
Just pushed an update to this that moves the implementation closer to the public interface. For tests, I opted to try to accomplish something slightly closer to a "unit" test since the expectations are not really being made about the exposed interfaces.

To be completely honest, I'm not thrilled about not having a test in place that exercises the problem itself from the user's standpoint, but I think we agree that changing the client_test to use another adapter at this point feels like overkill.

Let me know what you think.

iamvery commented:
I think I've been screwing up the terminology in talking about this problem. The thing I'm referring to is not a channel as much as it is the name of the pubsub queue. Am I understanding that correctly now? It doesn't really change anything, but I wanted to point that out...

dhh commented:
👍

iamvery commented:
❤️
